Update to Apache Kerby 1.1.1

dwalluck opened the pull request.

dwalluck commented:
There is a regression in the `KrbLoginResponseInfo` parsing. I have added tests which are currently failing.

Also, what do you think about making this dependency optional and looking for the class dynamically? Hopefully, we can move to Kerberos over HTTP in the future as we already have the Apache HttpComponents dependency via jhttpc.

dwalluck commented on pom.xml:
```diff
@@ -47,7 +47,7 @@
         <atlasVersion>0.16.0</atlasVersion>
         <mockitoVersion>2.7.20</mockitoVersion>
-        <kerbyVersion>1.0.1</kerbyVersion>
+        <kerbyVersion>1.1.1</kerbyVersion>
```
This fixes some bugs (so we can remove some local changes) and removes a transitive dependency (which we didn't use) that was insecure. I am trying to submit PRs upstream for the other changes, so we can eventually remove all local changes.

ruhan1 commented on src/test/resources/messages/krbLogin-request.xml:
```diff
@@ -0,0 +1,20 @@
+dmVyIHRoZSBsYXp5IGRvZy4KVGhlIHF1aWNrIGJyb3duIGZveCBqdW1wcyBvdmVyIHRoZSBsYXp5
+IGRvZy4KVGhlIHF1aWNrIGJyb3duIGZveCBqdW1wcyBvdmVyIHRoZSBsYXp5IGRvZy4KVGhlIHF1
+aWNrIGJyb3duIGZveCBqdW1wcyBvdmVyIHRoZSBsYXp5IGRvZy4=
+</string>
```
Can you try to remove the last \n before </string>? It seems STAX thinks there are two CHAR events. With the first event, it read the whole string until =; with the second, it reads a "" and override the prev value.

dwalluck replied:
This changes things a bit. Now we seem to get the portion of the string after the last `\n` instead of an empty string in some cases. In any case, it looks like the final `String` value is just the last `CHAR` event that happens instead of the entire text between the `<string></string>` tags, if what you say is true.

However, it looks like you have found that my tests are wrong. The expected string must end in a newline according to the spec (and this is also what the server sends), so the XML files are correct. I will update the expected test strings to end in a newline (by the way, the newline characters may vary depending on which operating system the client and server are running), but I stuck to UNIX in both cases for simplicity.

dwalluck commented:
> Expected strings should always end with a newline

To be clear, we still aren't getting the full string in either case.

At least you made me realize that while the text in the XML file had a final newline, the expected string in the tests did not. The XML and the tests should line up now, whereas the original tests would have failed even if the strings were parsing properly.

dwalluck commented:
It looks like it needs to be concatenating this data, and, as a general note, a string can be used to encode binary data, so we can't be trimming the string data itself:

> Any characters are allowed in a string except < and &, which are encoded as &lt; and &amp;. A string can be used to encode binary data.

It seems that Koji could have used the `<base64>` tag for base64-encoded binary data as opposed to the `<string>` tag.

dwalluck commented:
> It seems that Koji could have used the <base64> tag for base64-encoded binary data as opposed to the <string> tag.

In that case, I wonder if I should just add a converter to kojiji to return the decoded data. This seems nicer than trying to decode it later on.

ruhan1 commented:
I think we can concatenate the text for string and base64. We need a fix to RWX. I will take care of it soon.

dwalluck commented:
> I think we can concatenate the text for string and base64. We need a fix to RWX. I will take care of it soon.

It wasn't clear to me if `<base64>` was supposed to contain newlines or not, but I suppose it could. However, it doesn't look like code can be shared  between the base64 decoding code and a base64 string converter. I don't thibk that RWX has its own converters. I think it makes more sense to have any converters to/from standard Java library types in RWX, and have only Koji-specific converters in Kojiji.

It seems that an XML-RPC string is really a Java `byte[]` and that an XML-RPC base64 is a (US-ASCII/UTF8-encoded) Java `String`. How is RWX handling strings vs. byte arrays? In any case, it's only a couple lines of code to decode base64 `string` to `byte[]`, but I am wondering if you annotate a byte[] Java field and it gets an XML-RPC string, will it be able to handle it?

ruhan1 commented:
For <base64>, RWX will convert it to and from a byte[]. Ref to https://github.com/Commonjava/RWX/blob/master/rwx/src/main/java/org/commonjava/rwx/vocab/ValueType.java#L252 . I think if you define a field as byte[], and it will be rendered to <base64>, and vice verse. 
But I really do not remember I tested it.

dwalluck commented:
> For , RWX will convert it to and from a byte[]. Ref to https://github.com/Commonjava/RWX/blob/master/rwx/src/main/java/org/commonjava/rwx/vocab/ValueType.java#L252 . I think if you define a field as byte[], and it will be rendered to , and vice verse.
> But I really do not remember I tested it.

OK, I will update this pull request. I am going to try to annotate a `byte[]` field with a custom converter that decodes the received base64 string directly into the `byte[]` in kojiji rather than have the application code have do it.

This is what RWX already does for a `base64` type at https://github.com/Commonjava/RWX/blob/master/rwx/src/main/java/org/commonjava/rwx/vocab/ValueType.java#L262, but I guess we must repeat the code like this.

ruhan1 commented:
The concatenation fix is https://github.com/Commonjava/RWX/pull/35 . Please review.

dwalluck commented:
> The concatenation fix is Commonjava/RWX#35 . Please review.

There are some problems here. After fixing a few tests that failed due to known whitespace, we have `src/test/resources/messages/getTaskRequest-response.xml` causing a `NullPointerException` and we have `src/test/resources/messages/krbLogin-response.xml` also failing parsing.

Please run the Kojiji tests against the RWX master.

dwalluck commented:
retest this please

ruhan1 commented:
Pls update master. A fix is done https://github.com/release-engineering/kojiji/pull/107

dwalluck commented:
> Pls update master. A fix is done #107

I actually removed the whitespace on this branch already and there were still three test failures. I think that the test results should be current.

dwalluck commented:
> Fix KrbLoginResponseTest

It turns out that it was an error in the data types and the order of the fields in the test itself. This constituted 2 of the 3 failures.

The one failure that I am still worried about is the `NullPointerException`. Since whitespace in a string is perfectly legal, the parser should never throw an exception on such input.

We just would not expect that a string with whitespace is identical to a string without (meaning, we would only trim a string if it were absolutely necessary).

dwalluck commented:
The tests are passing, but it's not ready to merge yet. The live login is failing for some other reason.

ruhan1 commented:
RWX has no problem handling whitespace in a string. As you said, this is a legal and common case. I can verify it works by a simple RWX unit test. But I did not test your case. Is there still problem with whitespace?

dwalluck commented:
> RWX has no problem handling whitespace in a string. As you said, this is a legal and common case. I can verify it works by a simple RWX unit test. But I did not test your case. Is there still problem with whitespace?

Why did the tests show a `NullPointerException` until you removed the whitespace from the `<string>` in the XML file? Or is that not what happened?

ruhan1 commented:
I am confused. Which NullPointerException? Do you mean I have removed something and fixed it?

dwalluck commented:
> I am confused. Which NullPointerException? Do you mean I have removed something and fixed it?

Specifically, https://nos-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/kojiji-PR-re-proj/com.redhat.red.build$kojiji/23/testReport/junit/com.redhat.red.build.koji.model.xmlrpc.messages/GetTaskRequestResponseTest/verifyVsCapturedHttp_asBuildRequest/

ruhan1 commented:
I only saw assert error but not NPE when I fixed the getTaskRequest-response.xml. And I can not make it throw NPE now by adding back the newlines surrounding the strings. Do you know how to reproduce it?

dwalluck commented:
> I only saw assert error but not NPE when I fixed the getTaskRequest-response.xml. And I can not make it throw NPE now by adding back the newlines surrounding the strings. Do you know how to reproduce it?

Sorry, I keep saying `NPE`, but I just mean `null`. How did a `null` value become non-null once you removed the whitespace from the string?

ruhan1 commented:
BTW, do you mean the whitespace in the middle of a string or at the beginning/end of a string when saying "whitespace in a string"? Before the concatenation fix, RWX trims the string value. Now it keeps all chars, newline, whitespace, wherever in the middle or beginning/end.

dwalluck commented:
I mean any whitespace between the `<string></string>` tags. If I remember correctly, it was something like
```
<string>
    text
</string>
```

and something was `null`, then it was changed to

```
<string>text</string>
```

and the `null` went away. Is that right?

dwalluck commented:
I don't remember the word for it... when the XML tags *hug* the text (meaning, there is no whitespace between the tags and the value itself, it seemed to make the test pass all of a sudden. At least I can't remember any code in that test changing, so it was just the whitespace in XML file in that case that caused the test to fail.

ruhan1 commented:
This is weird. I did not see a null value during my test. I guess I saw something like "\n text \n", then I removed the newlines.

dwalluck commented:
I believe that test was looking at a `List<String>` that had some `null` elements in it. But, it's hard to see how the some string becomes `null` if the whitespace is not being trimmed. Is it definitely the case that lines with only `\n` in them stay as `\n` and do not become `null`?

dwalluck commented:
> This is weird. I did not see a null value during my test. I guess I saw something like "\n text \n", then I removed the newlines.

All I can think is for that test, to revert to the previous file with the whitespace and make sure it does not have some `null`. If it passes, I don't have any other ideas right now.

ruhan1 commented:
Yeah. I create another PR. https://github.com/release-engineering/kojiji/pull/109

dwalluck commented:
> Yeah. I create another PR. #109

Looks like it passed. So, do we have any idea why it failed before?

ruhan1 commented:
No idea. The null is weird. Maybe that failure used old RWX code without the concatenate fix?

ruhan1 commented:
Forget about it. If it used old RWX code, it passed always. .. anyway, no idea.

dwalluck commented:
> Forget about it. If it used old RWX code, it passed always. .. anyway, no idea.

Yes, I agree. 

Also, it seems that even the full test log does not show anything being downloaded, so I was unable to determine the RWX version from that. In general, I don't know how to find the dependency versions being used for these test builds.

ruhan1 commented:
Okay. I will merge #109. Forget about the null until it hits us again (I suppose it won't)

dwalluck commented:
It should be OK to merge now.
